arrows/gdal/image_io.cxx

```diff
@@ -0,0 +1,250 @@
+ */
+
+/**
+ * \file
+ * \brief GDAL image_io implementation
+ */
+
+#include "image_io.h"
+
+#include <vital/types/metadata.h>
+#include <vital/types/metadata_traits.h>
+
+#include <gdal_priv.h>
+
+#include <sstream>
+
+namespace kwiver {
+namespace arrows {
+namespace gdal {
+
+void add_metadata(char* raw_md, vital::metadata_sptr md)
+{
+  std::istringstream md_string(raw_md);
+
+  // Get the key
+  std::string key;
+  if ( !std::getline( md_string, key, '=') )
+  {
+    return;
+  }
+
+  // Get the value
+  std::string value;
+  if ( !std::getline( md_string, value, '=') )
+  {
+    return;
+  }
+
+#define MAP_METADATA_SCALAR( GN, KN )                       \
+if ( key == #GN)                                            \
+{                                                           \
+  md->add( NEW_METADATA_ITEM(                               \
+    vital::VITAL_META_RPC_ ## KN, std::stod( value ) ) ) ;  \
+}                                                           \
+
+#define MAP_METADATA_COEFF( GN, KN )          \
+if ( key == #GN)                              \
+{                                             \
+  md->add( NEW_METADATA_ITEM(                 \
+    vital::VITAL_META_RPC_ ## KN, value ) );  \
+}                                             \
+
+  MAP_METADATA_SCALAR( HEIGHT_OFF,   HEIGHT_OFFSET )
+  MAP_METADATA_SCALAR( HEIGHT_SCALE, HEIGHT_SCALE )
+  MAP_METADATA_SCALAR( LONG_OFF,     LONG_OFFSET )
+  MAP_METADATA_SCALAR( LONG_SCALE,   LONG_SCALE )
+  MAP_METADATA_SCALAR( LAT_OFF,      LAT_OFFSET )
+  MAP_METADATA_SCALAR( LAT_SCALE,    LAT_SCALE )
+  MAP_METADATA_SCALAR( LINE_OFF,     ROW_OFFSET )
+  MAP_METADATA_SCALAR( LINE_SCALE,   ROW_SCALE )
+  MAP_METADATA_SCALAR( SAMP_OFF,     COL_OFFSET )
+  MAP_METADATA_SCALAR( SAMP_SCALE,   COL_SCALE )
+
+  MAP_METADATA_COEFF( LINE_NUM_COEFF, ROW_NUM_COEFF )
+  MAP_METADATA_COEFF( LINE_DEN_COEFF, ROW_DEN_COEFF )
+  MAP_METADATA_COEFF( SAMP_NUM_COEFF, COL_NUM_COEFF )
+  MAP_METADATA_COEFF( SAMP_DEN_COEFF, COL_DEN_COEFF )
+
+#undef MAP_METADATA_SCALAR
+#undef MAP_METADATA_COEFF
+}
+
+/// Private implementation class
+class image_io::priv
+{
+public:
+  /// Constructor
+  priv() : m_logger( vital::get_logger( "arrows.gdal.image_io" ) ) {}
+
+  vital::logger_handle_t m_logger;
+};
+
+/// Constructor
+image_io
+::image_io()
+: d_(new priv)
+{
+}
+
+/// Destructor
+image_io
+::~image_io()
+{
+}
+
+/// Load image image from the file
+/**
+ * \param filename the path to the file the load
+ * \returns an image container refering to the loaded image
+ */
+vital::image_container_sptr
+image_io
+::load_(const std::string& filename) const
+{
+  GDALAllRegister();
+
+  std::shared_ptr<GDALDataset> gdalDataset;
+
+  gdalDataset.reset( static_cast<GDALDataset*>(
+    GDALOpen( filename.c_str(), GA_ReadOnly ) ) );
+
+  vital::image img;
+
+  if ( !gdalDataset )
+  {
+    LOG_ERROR( d_->m_logger, "Unable to load data from " << filename );
+    throw vital::image_exception();
+  }
```

Comment:
All the error messages are now passed to the exceptions rather than sent to the logger. I switched to using the VITAL_THROW macro since that seems to provide addition information with a similar amount of code.